### ISSUE TITLE
[QUESTION] Behaviour of <a-a> unclear from documentation

### ISSUE THREAD

**Issue by ismay** [mawww/kakoune#4570]
### Question

First off, thanks for a great editor and all the work you've done on it! One thing is a little unclear to me from the docs. Looking at the docs for whole object selection [here](https://github.com/mawww/kakoune/blob/master/doc/pages/keys.asciidoc#whole-object), it states "select the whole object" for <kbd>Alt</kbd> + <kbd>a</kbd>.

Now in most cases this is clear. But I encountered a scenario where the behaviour was unexpected for me as a new user, and the docs don't make it clear why the command behaves in this manner.

Let me explain: 

- Say I am on the last double quote, for a pair of double quotes:

![image](https://user-images.githubusercontent.com/7355199/158987086-68480b71-4d1a-42ad-b1f0-11b417577d2a.png)

- Now I press <kbd>Alt</kbd> + <kbd>A</kbd>, expecting it to select backwards to the `"` preceding it, but what actually happens is that it selects forward to a couple lines below:

![image](https://user-images.githubusercontent.com/7355199/158987306-4474092a-80aa-4080-b893-b0e6cc350d64.png)

- This does not happen when I'm on the first of the pair of double quotes, like so:

![image](https://user-images.githubusercontent.com/7355199/158987406-5d91bd21-7988-4300-bb19-8be31e051190.png)

- In that case it selects forwards, like I would expect it to:

![image](https://user-images.githubusercontent.com/7355199/158987452-69e24372-ff52-4f61-a3ed-a0a340e4aaed.png)

So to me, as a relatively new user, it is not clear why this is not working as I expected. The docs don't really explain this for me. Why doesn't <kbd>Alt</kbd> + <kbd>A</kbd> select backwards in the above scenario?

**Comment by Screwtapello** [comment]
Kakoune's algorithm is basically:

- if the cursor is not on a `"`, search forward and backward to find the enclosing `"` characters
- if the cursor *is* on a `"`, it could be the opening quote or the close quote — for consistency, just assume it's the opening quote and the next `"` encountered is the close quote

If the text object you're trying to select is delimited by an opening/closing pair of characters, Kakoune can be smarter about which direction to search for a match:

```
word (word here)
 something

something here: (word)
```

If you put the cursor on the `)` in the first line and press `<a-a>)`, then Kakoune does indeed select `(word here)` as you'd expect.

On the other hand,  there are situations where Kakoune's current behaviour is the expected one:

```
printf "'%s'\n" "
    info %{ hello world }
"
```

Put the cursor on the `"` at the end of the first line there and press `<a-a>"` and Kakoune selects the entire multi-line string.

In short, in a document with multiple pairs of `"`, there's no way for Kakoune to know which text is "inside" the quotes and which text is "outside" without a much deeper understanding of the file-format being edited. Kakoune's built-in editing primitives try to avoid understanding any individual file format, so they can be useful in as many formats as possible.

**Comment by ismay** [comment]
I see, yeah that makes sense. I hadn't considered the difference between pairs that distinguish between the opening and closing character (like `{}`) and ones that don't (like `""`).

> In short, in a document with multiple pairs of ", there's no way for Kakoune to know which text is "inside" the quotes and which text is "outside" without a much deeper understanding of the file-format being edited. Kakoune's built-in editing primitives try to avoid understanding any individual file format, so they can be useful in as many formats as possible.

Yeah that sounds like a solid approach. I prefer a simple, predictable algorithm over a smarter one that might fail in complex ways that are harder to adapt to. It was just a little unclear to me why this was happening (if I was using the wrong command, etc.), but now that I know it should be easy to adapt to!

Thanks for the explanation!
